src/OrleansRuntime/GrainDirectory/RemoteGrainDirectory.cs

```diff
@@ -22,313 +20,53 @@ internal RemoteGrainDirectory(LocalGrainDirectory r, GrainId id)
             logger = TraceLogger.GetLogger("Orleans.GrainDirectory.CacheValidator", TraceLogger.LoggerType.Runtime);
         }
 
-        public async Task<int> Register(ActivationAddress address, int retries)
+        public async Task<AddressAndTag> RegisterAsync(ActivationAddress address, bool singleActivation, int hopCount)
         {
-            router.RegistrationsRemoteReceived.Increment();
-            // validate that this grain should be stored in our partition
-            SiloAddress owner = router.CalculateTargetSilo(address.Grain);
-            if (owner == null)
-            {
-                // We don't know about any other silos, and we're stopping, so throw
-                throw new InvalidOperationException("Grain directory is stopping");
-            }
-            
-            if (router.MyAddress.Equals(owner))
-            {
-                router.RegistrationsLocal.Increment();
-                return partition.AddActivation(address.Grain, address.Activation, address.Silo);
-            }
-
-            if (retries > 0)
-            {
-                if (logger.IsVerbose2) logger.Verbose2("Retry " + retries + " RemoteGrainDirectory.Register for address=" + address + " at Owner=" + owner);
-                PrepareForRetry(retries);
-
-                await Task.Delay(RETRY_DELAY);
-                
-                SiloAddress o = router.CalculateTargetSilo(address.Grain);
-                if (o == null)
-                {
-                    // We don't know about any other silos, and we're stopping, so throw
-                    throw new InvalidOperationException("Grain directory is stopping");
-                }
-                if (router.MyAddress.Equals(o))
-                {
-                    router.RegistrationsLocal.Increment();
-                    return partition.AddActivation(address.Grain, address.Activation, address.Silo);
-                }
-                router.RegistrationsRemoteSent.Increment();
-                return await GetDirectoryReference(o).Register(address, retries - 1);
-            }
+            (singleActivation ? router.RegistrationsSingleActRemoteReceived : router.RegistrationsRemoteReceived).Increment();
             
-            throw new OrleansException("Silo " + router.MyAddress + " is not the owner of the grain " + address.Grain + " Owner=" + owner);
+            return await router.RegisterAsync(address, singleActivation, hopCount);
         }
 
-        public Task RegisterMany(List<ActivationAddress> addresses, int retries)
+        public Task RegisterMany(List<ActivationAddress> addresses, bool singleActivation)
         {
-            return Task.WhenAll(addresses.Select(addr => Register(addr, retries)));
-        }
+            if (addresses == null || addresses.Count == 0)
+                throw new ArgumentException("addresses");
```

Comment:
A nit: `ArgumentException` takes an error message as a constructor argument, e.g. "addresses cannot be an empty list or null" and an optional parameter name. `ArgumentNullException` OTOH takes just an argument name.